Add feature to run Geb specs with Testcontainers

matrei opened the pull request:
This PR add a `ContainerGebSpec` class that can be used in place of `GebSpec` and that will run the test browser in a container instead of directly on the host machine.

matrei committed:
feat: Add `ContainerGebSpec` for running the browser in a container

This feature enables easy running of geb tests in a container so you don't have to manage driver versions on the host machine.

matrei committed:
fix: Set proper package for `ContainerGebSpec`

The common package name used by grails plugins is `grails.plugin.{plugin-name}`.

matrei committed:
refactor: Move `GebGrailsPlugin` to package `grails.plugin.geb`

The common package name used by grails plugins is `grails.plugin.{plugin-name}`.

matrei committed:
fix: Throw `IllegalStateException` if `serverPort` is null

For `ContainerGebSpec` to work, the serverPort needs to be set.
This is done by adding the `@Integration` annotation to the test class.

matrei committed:
fix: Exclude unnecessary slf4j exclusion in the pom

matrei committed:
fix: Major updates to `ContainerGebSpec` feature

- Use Gradle testFixtures feature to make test dependencies available to downstream projects without leaking them to the wrong scopes
- Refactor `ContainerGebSpec` in co-operation with sbglasius to add more features and make it work without Docker Desktop.

matrei commented:
> I am apparently not committer on grails

Well, you should be. That is an easy fix 😄

jamesfredley commented:
@sbglasius I have submitted a request and you should see invites to a github team soon.

jamesfredley reviewed (approved):
I suspect the servlet-api changes will likely change based on https://github.com/gpc/grails-mail/pull/56#issuecomment-2415877468

matrei committed:
fix: Change scope of dependencies in generated tests

After this change this library needs to be referenced with `integrationTestImplementation testFixtures('org.grails.plugins:geb')`
but now we are not leaking any test dependencies to downstream projects non-test configurations.

matrei committed:
fix: Remove `compileOnly` dependency on `servlet-api`

matrei committed:
chore: Move package declaration in `ContainerGebSpec`

jamesfredley commented:
https://github.com/grails/grails-core/pull/13763

matrei committed:
fix(deps): Use versions from the `grails-bom` where available

matrei committed:
build: Update to `develocity` plugin and `bellsoft` toolchain vendor

jamesfredley commented on build.gradle:
```diff
@@ -19,6 +19,7 @@ apply plugin: 'maven-publish'
 java {
     toolchain {
         languageVersion = JavaLanguageVersion.of(17)
+        vendor = JvmVendorSpec.BELLSOFT
```
If we add this we will also want to add a plugin to settings.gradle.  See https://github.com/grails/grails-core/issues/13783

matrei replied:
I think I did that. Is there anything missing from settings.gradle?

jamesfredley replied:
Sorry, I must have been looking at a stale version of the PR.  Looks great.

matrei committed:
chore(build): Remove unused project properties

matrei committed:
fix(deps): Use versions from `grails-bom`

matrei committed:
docs: Update readme with info about `ContainerGebSpec`
